File: engines/unity/android/src/main/kotlin/com/xraph/gameframework/unity/UnityEnginePlugin.kt
package com.xraph.gameframework.unity

import android.content.Context
import androidx.lifecycle.Lifecycle
import com.xraph.gameframework.gameframework.core.GameEngineController
import com.xraph.gameframework.gameframework.core.GameEngineFactory
import com.xraph.gameframework.gameframework.core.GameEngineRegistry
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.BinaryMessenger

/**
 * UnityEnginePlugin - Plugin for Unity Engine integration
 *
 * This plugin registers the Unity engine factory with the game framework,
 * allowing Unity engines to be embedded in Flutter applications.
 */
class UnityEnginePlugin : FlutterPlugin {

    companion object {
        private const val ENGINE_TYPE = "unity"

        /**
         * Manual registration method for early initialization
         *
         * Call this before the Flutter engine is fully initialized if needed.
         */
        @JvmStatic
        fun register() {
            GameEngineRegistry.instance.registerFactory(
                ENGINE_TYPE,
                UnityEngineFactory::class.java.newInstance()
            )
        }
    }

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        // Register Unity factory with the game framework
        val factory = UnityEngineFactory(binding.binaryMessenger)
        GameEngineRegistry.instance.registerFactory(ENGINE_TYPE, factory)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        // Unregister Unity factory
        GameEngineRegistry.instance.unregisterFactory(ENGINE_TYPE)
    }
}

/**
 * Factory for creating Unity engine controllers
 */
class UnityEngineFactory(
    messenger: BinaryMessenger
) : GameEngineFactory(messenger) {

    override fun createController(
        context: Context,
        viewId: Int,
        messenger: BinaryMessenger,
        lifecycle: Lifecycle,
        config: Map<String, Any?>
    ): GameEngineController {
        return UnityEngineController(
            context,
            viewId,
            messenger,
            lifecycle,
            config
        )
    }
}

File: engines/unity/dart/ios/Classes/UnityEnginePlugin.swift
import Flutter
import Foundation
import gameframework

/**
 * UnityEnginePlugin - Plugin for Unity Engine integration
 *
 * This plugin registers the Unity engine factory with the game framework,
 * allowing Unity engines to be embedded in Flutter applications on iOS.
 */
public class UnityEnginePlugin: NSObject, FlutterPlugin {

    private static let engineType = "unity"

    public static func register(with registrar: FlutterPluginRegistrar) {
        // Auto-initialize Unity integration with command-line arguments
        // This eliminates the need for manual AppDelegate modification
        // Note: launchOptions is not available here, but Unity works without it for most use cases
        UnityPlayerUtils.shared.InitUnityIntegrationWithOptions(
            argc: CommandLine.argc,
            argv: CommandLine.unsafeArgv,
            nil
        )
        NSLog("UnityEnginePlugin: Auto-initialized Unity integration")
        
        // Register Unity factory with the game framework
        let factory = UnityEngineFactory()
        GameEngineRegistry.shared.registerFactory(
            engineType: engineType,
            factory: factory
        )
        
        // Trigger platform view registration (lazy registration pattern)
        GameframeworkPlugin.registerPlatformView(engineType: engineType)
    }

    /**
     * Manual registration method for early initialization
     *
     * Call this before the Flutter engine is fully initialized if needed.
     */
    public static func registerManually() {
        let factory = UnityEngineFactory()
        GameEngineRegistry.shared.registerFactory(
            engineType: engineType,
            factory: factory
        )
    }
}

/**
 * Factory for creating Unity engine controllers
 */
public class UnityEngineFactory: NSObject, GameEngineFactory {

    public func createController(
        frame: CGRect,
        viewId: Int64,
        messenger: FlutterBinaryMessenger,
        config: [String: Any]
    ) -> GameEnginePlatformView {
        return UnityEngineController(
            frame: frame,
            viewId: viewId,
            messenger: messenger,
            config: config
        )
    }
}
